### DISCUSSION TITLE
SVG stroke width difference on linux (ubuntu) vs. macos - totally puzzling.

### DISCUSSION THREAD

**Discussion by noahlh** [ImageMagick/ImageMagick#5166] [Help]
Hello!  I have been banging my head against the wall on this one for the past day, so I am appealing to the wisdom of the crowd for some advice.

The core issue:  The exact same SVG is rendering with thicker strokes when processed on a linux machine (running in Google Cloud) vs. my local Mac.

I'm converting some SVGs to PNGs in an application I'm building, and am primarily interfacing with ImageMagick via the C bindings.  But I've been able to reproduce this on the command line as well.

The input SVG is of the format...

```
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="171.7484px" height="25px">
<path fill="#000" d="M4.4981,23.5224l..........>
</path>
</svg>
```

(In other words, there's no stroke tag or any other attributes other than one big path tag...)

I'm using a command like this:

`magick convert -density 300 input.svg output.png`

Here's what the SVG looks like in a browser:
<img width="166" alt="Screen Shot 2022-05-19 at 7 03 20 PM" src="https://user-images.githubusercontent.com/113276/169433906-21763079-0d49-4f6e-9168-6a868e5d5036.png">

Here's what the output PNG looks like running the above command on my mac:
![test-mac](https://user-images.githubusercontent.com/113276/169434002-a7e46b15-d8b7-4fa2-8d5d-7955970f2508.png)

And here's the output PNG of the exact same input SVG running running the exact same command on Ubuntu 20.04 in Google Cloud:
![test](https://user-images.githubusercontent.com/113276/169434115-5823b577-9c87-4c22-8a1f-0ec03d63d956.png)

When I first started investigating, I discovered the Ubuntu instance was running ImageMagick 6.9 and thought ah ha!  That must be it.  But even when I upgraded to 7.1.0 the results are the same.

Output of `magick -version` on my local Mac:

```
Version: ImageMagick 7.1.0-29 Q16-HDRI arm 19841 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0) 
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
Compiler: gcc (4.2)
```

Output on the Ubuntu instance:

```
Version: ImageMagick 7.1.0-35 Q16-HDRI x86_64 20062 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): freetype jng jpeg png x xml zlib
Compiler: gcc (9.4)
```

I also ran `convert -list format | grep -i svg` on each and got this output:

Local mac:
```
     MSVG* SVG       rw+   ImageMagick's own SVG internal renderer
      SVG* SVG       rw+   Scalable Vector Graphics (XML 2.9.4)
     SVGZ* SVG       rw+   Compressed Scalable Vector Graphics (XML 2.9.4)
```

Ubuntu:
```
     MSVG* rw+   ImageMagick's own SVG internal renderer
      SVG* rw+   Scalable Vector Graphics (XML 2.9.10)
     SVGZ* rw+   Compressed Scalable Vector Graphics (XML 2.9.10)
```

I also tried outputting in JPEG and the results are the same.  So it's something to do with the input SVG parsing/rendering (I think), as opposed to my PNG library.

I know there are some subtle version differences up there.  But before I spend the next week ripping apart and upgrading/downgrading every library, I'm hoping someone can point me in the right direction with what's going on.

Thanks in advance!!

**Comment by snibgo** [comment]
IM delegates rasterization of SVG to either the internal MSVG, or RSVG, or Inkscape. Of these three, Inkscape is the most complete (but also the slowest). If Inkscape is on your system path, IM will use it.

I suspect your two systems use two different rasterizers. At the command line, insert "-verbose" as the first argument, like this:
```
magick convert -verbose -density 300 input.svg output.png
```
The text output will probably say which rasterizer is being used.

**Reply by noahlh** [reply]
@snibgo Thanks!  Both are using inkscape:

Local Mac:
```
'inkscape' '/var/folders/8d/2160q8z14373ks1df31w6tx40000gn/T/magick-kyvC2Gg-dOb4kdutBpM8nWVxXlUneEVM' --export-filename='/var/folders/8d/2160q8z14373ks1df31w6tx40000gn/T/magick-QmfnOXiQz8RuRUTDCBZzqOw2p1GgLhMt.png' --export-dpi='300' --export-background='rgb(100%,100%,100%)' --export-background-opacity='0.99999999999900002212' > '/var/folders/8d/2160q8z14373ks1df31w6tx40000gn/T/magick-LfAIlc1-1dyW2ZHaSJlUyO3-7k2txPz4' 2>&1
mvg:/var/folders/8d/2160q8z14373ks1df31w6tx40000gn/T/magick-vvScM1vh0CB_Mr2tsFIowWK-g2oz7WYK=>/var/folders/8d/2160q8z14373ks1df31w6tx40000gn/T/magick-vvScM1vh0CB_Mr2tsFIowWK-g2oz7WYK MVG 537x78 537x78+0+0 16-bit sRGB 7747B 0.020u 0:00.015
test.svg SVG 537x78 537x78+0+0 16-bit sRGB 7747B 0.000u 0:00.000
test.svg=>test.png SVG 537x78 537x78+0+0 16-bit sRGB 7747B 0.000u 0:00.005
```


Ubuntu:
```
'inkscape' '/tmp/magick-t-3pREE58jUQ2W8hErV29QugtWGQ9Ro9' --export-filename='/tmp/magick-PQ9qvvuqnxdzziGJiByXfzpqy8oI0eV8.png' --export-dpi='300' --export-background='rgb(100%,100%,100%)' --export-background-opacity='0.99999999999900002212' > '/tmp/magick-GOm5HiMB1u5WRAEdQcRT885eVId8Ecpa' 2>&1
mvg:/tmp/magick-XjRT2sPRDeyUn-jQJBy2pRc_-tCVFip8=>/tmp/magick-XjRT2sPRDeyUn-jQJBy2pRc_-tCVFip8 MVG 537x78 537x78+0+0 16-bit sRGB 7370B 0.100u 0:00.096
test.svg SVG 537x78 537x78+0+0 16-bit sRGB 7370B 0.000u 0:00.000
test.svg=>test.png SVG 537x78 537x78+0+0 16-bit sRGB 7370B 0.000u 0:00.008
```

But this gives me an avenue to explore.  Maybe there's a subtle version difference there.

**Comment by noahlh** [comment]
Ok @snibgo you definitely pointed me in the right direction.  Upgrading the Ubuntu server to IM version 7.1.0 and Inkscape version 1.1 seems to be getting me the correct output via command-line.  Progress!

Next:  

The code I'm writing calls the C API (specifically, `MagickSetResolution`, `MagickReadImageBlob` (to get the SVG in), `MagickSetImageFormat` (to set the output to PNG), and then `MagickGetImageBlob` to get the final PNG and write it to IO.

I compiled and ran again on the same system where the correct versions are now installed, but the problem persists there.  

Is there any way to force the C API to use the proper Inkscape version?  Or does it default to the internal SVG renderer?

**Comment by snibgo** [comment]
I expect the function MagickReadImageBlob() will read an SVG in exactly the same way as a command-line execution of "magick".

**Comment by noahlh** [comment]
Ok an update for anyone who comes across this:  I had to admit defeat on this particular bug -- I just couldn't figure out where the alternate inkscape was being called from and ended up ripping out ImageMagick altogether :(  Since it was just calling inkscape to do the SVG to PNG conversion, I ended up replacing the ImageMagick API code with a direct shell call to inkscape (which I installed separately - version 1.1), and that seems to be converting the SVGs properly without the weird extra thickness.
